File: algorithms/easy/maximum_units_on_a_truck.rs
impl Solution {
    pub fn maximum_units(mut box_types: Vec<Vec<i32>>, truck_size: i32) -> i32 {
        box_types.sort_by(|a, b| b[1].cmp(&a[1]));
        let mut total = 0;
        let mut ans = 0;
        for each in &box_types {
            if total >= truck_size {
                break;
            }
            let n = each[0];
            let x = each[1];
            if total + n <= truck_size {
                total += n;
                ans += n * x;
            } else {
                ans += (truck_size - total) * x;
                total = truck_size;
            }
        }
        ans
    }
}

File: algorithms/easy/maximum_units_on_a_truck.ex
defmodule Solution do
  @spec maximum_units(box_types :: [[integer]], truck_size :: integer) :: integer
  def maximum_units(box_types, truck_size) do
    box_types
    |> Enum.sort_by(fn [_, k] -> -k end)
    |> Enum.reduce_while({0, 0}, fn [n, x], {total, ans} ->
      cond do
        total >= truck_size -> {:halt, {total, ans}}
        total + n <= truck_size -> {:cont, {total + n, ans + n * x}}
        true -> {:cont, {truck_size, ans + (truck_size - total) * x}}
      end
    end)
    |> elem(1)
  end
end
